File: clients/reactive-ts/src/sync-lifecycle.ts
/** Formally modeled mobile/desktop ownership lifecycle. */
export type SyncLifecyclePhase =
  | 'idle'
  | 'acquiring'
  | 'running'
  | 'releasing'
  | 'closed';

export type SyncLifecycleEvent =
  | 'wake'
  | 'join'
  | 'begin-acquire'
  | 'acquire-granted'
  | 'acquire-deferred'
  | 'cancel'
  | 'cycle-settled'
  | 'release-settled'
  | 'close'
  | 'process-abort';

export interface SyncLifecycleSnapshot {
  readonly phase: SyncLifecyclePhase;
  readonly wakePending: boolean;
  readonly closeRequested: boolean;
  readonly cancelRequested: boolean;
  readonly permitHeld: boolean;
}

export const initialSyncLifecycle: SyncLifecycleSnapshot = Object.freeze({
  phase: 'idle',
  wakePending: false,
  closeRequested: false,
  cancelRequested: false,
  permitHeld: false,
});

export function isValidSyncLifecycle(state: SyncLifecycleSnapshot): boolean {
  const activePermit = state.phase === 'running' || state.phase === 'releasing';
  if (state.permitHeld !== activePermit) return false;
  if (state.phase === 'closed') {
    return (
      state.closeRequested &&
      !state.wakePending &&
      !state.cancelRequested &&
      !state.permitHeld
    );
  }
  if (state.closeRequested && state.wakePending) return false;
  return !state.cancelRequested || state.phase === 'running';
}

function snapshot(
  state: SyncLifecycleSnapshot,
  changes: Partial<SyncLifecycleSnapshot>,
): SyncLifecycleSnapshot {
  return Object.freeze({ ...state, ...changes });
}

/** Pure transition relation mirrored by `formal/mobile_desktop_lifecycle.qnt`. */
export function transitionSyncLifecycle(
  state: SyncLifecycleSnapshot,
  event: SyncLifecycleEvent,
): SyncLifecycleSnapshot | undefined {
  if (!isValidSyncLifecycle(state)) return undefined;
  let next: SyncLifecycleSnapshot | undefined;
  switch (event) {
    case 'wake':
      if (state.phase === 'closed' || state.closeRequested) return undefined;
      next = snapshot(state, { wakePending: true });
      break;
    case 'join':
      if (
        state.phase !== 'acquiring' &&
        state.phase !== 'running' &&
        state.phase !== 'releasing'
      ) {
        return undefined;
      }
      next = state;
      break;
    case 'begin-acquire':
      if (
        state.phase !== 'idle' ||
        !state.wakePending ||
        state.closeRequested
      ) {
        return undefined;
      }
      next = snapshot(state, { phase: 'acquiring', wakePending: false });
      break;
    case 'acquire-granted':
      if (state.phase !== 'acquiring') return undefined;
      next = snapshot(state, {
        phase: state.closeRequested ? 'releasing' : 'running',
        permitHeld: true,
        cancelRequested: false,
      });
      break;
    case 'acquire-deferred':
      if (state.phase !== 'acquiring') return undefined;
      next = snapshot(state, {
        phase: state.closeRequested ? 'closed' : 'idle',
        wakePending: state.closeRequested ? false : state.wakePending,
        cancelRequested: false,
        permitHeld: false,
      });
      break;
    case 'cancel':
      if (state.phase !== 'running') return undefined;
      next = snapshot(state, { cancelRequested: true });
      break;
    case 'cycle-settled':
      if (state.phase !== 'running' || !state.permitHeld) return undefined;
      next = snapshot(state, {
        phase: 'releasing',
        cancelRequested: false,
      });
      break;
    case 'release-settled':
      if (state.phase !== 'releasing' || !state.permitHeld) return undefined;
      next = snapshot(state, {
        phase: state.closeRequested ? 'closed' : 'idle',
        wakePending: state.closeRequested ? false : state.wakePending,
        cancelRequested: false,
        permitHeld: false,
      });
      break;
    case 'close':
      if (state.phase === 'closed') return undefined;
      if (state.phase === 'idle') {
        next = snapshot(state, {
          phase: 'closed',
          wakePending: false,
          closeRequested: true,
          cancelRequested: false,
        });
      } else {
        next = snapshot(state, {
          wakePending: false,
          closeRequested: true,
          cancelRequested: state.phase === 'running',
        });
      }
      break;
    case 'process-abort':
      if (
        state.phase !== 'acquiring' &&
        state.phase !== 'running' &&
        state.phase !== 'releasing'
      ) {
        return undefined;
      }
      next = snapshot(state, {
        phase: state.closeRequested ? 'closed' : 'idle',
        wakePending: false,
        cancelRequested: false,
        permitHeld: false,
      });
      break;
  }
  return next && isValidSyncLifecycle(next) ? next : undefined;
}

export class SyncLifecycleTransitionError extends Error {
  readonly before: SyncLifecycleSnapshot;
  readonly event: SyncLifecycleEvent;

  constructor(
    before: SyncLifecycleSnapshot,
    event: SyncLifecycleEvent,
  ) {
    super(
      `undefined opto-sync lifecycle transition: ${before.phase} + ${event}`,
    );
    this.name = 'SyncLifecycleTransitionError';
    this.before = before;
    this.event = event;
  }
}

/** Production state machine; undefined events fail closed without mutation. */
export class SyncLifecycleMachine {
  #state = initialSyncLifecycle;

  get state(): SyncLifecycleSnapshot {
    return this.#state;
  }

  apply(event: SyncLifecycleEvent): SyncLifecycleSnapshot {
    const next = transitionSyncLifecycle(this.#state, event);
    if (!next) throw new SyncLifecycleTransitionError(this.#state, event);
    this.#state = next;
    return next;
  }
}

File: clients/reactive-dart/lib/src/sync_lifecycle.dart
/// Formally modeled lifecycle shared by mobile and desktop sync runners.
///
/// The transition relation is specified in
/// `formal/mobile_desktop_lifecycle.qnt`. Runners must use [apply] for every
/// ownership change so an implementation drift fails closed as a
/// [SyncLifecycleTransitionError] instead of creating an unmodeled state.
enum SyncLifecyclePhase { idle, acquiring, running, releasing, closed }

enum SyncLifecycleEvent {
  wake,
  join,
  beginAcquire,
  acquireGranted,
  acquireDeferred,
  cancel,
  cycleSettled,
  releaseSettled,
  close,
  processAbort,
}

final class SyncLifecycleSnapshot {
  const SyncLifecycleSnapshot({
    required this.phase,
    required this.wakePending,
    required this.closeRequested,
    required this.cancelRequested,
    required this.permitHeld,
  });

  const SyncLifecycleSnapshot.initial()
    : phase = SyncLifecyclePhase.idle,
      wakePending = false,
      closeRequested = false,
      cancelRequested = false,
      permitHeld = false;

  final SyncLifecyclePhase phase;
  final bool wakePending;
  final bool closeRequested;
  final bool cancelRequested;

  /// A local execution permit, backed by a durable fence on desktop.
  final bool permitHeld;

  bool get isValid {
    final activePermit =
        phase == SyncLifecyclePhase.running ||
        phase == SyncLifecyclePhase.releasing;
    if (permitHeld != activePermit) return false;
    if (phase == SyncLifecyclePhase.closed) {
      return closeRequested && !wakePending && !cancelRequested && !permitHeld;
    }
    if (closeRequested && wakePending) return false;
    if (cancelRequested && phase != SyncLifecyclePhase.running) return false;
    return true;
  }

  SyncLifecycleSnapshot copyWith({
    SyncLifecyclePhase? phase,
    bool? wakePending,
    bool? closeRequested,
    bool? cancelRequested,
    bool? permitHeld,
  }) => SyncLifecycleSnapshot(
    phase: phase ?? this.phase,
    wakePending: wakePending ?? this.wakePending,
    closeRequested: closeRequested ?? this.closeRequested,
    cancelRequested: cancelRequested ?? this.cancelRequested,
    permitHeld: permitHeld ?? this.permitHeld,
  );

  @override
  String toString() =>
      'SyncLifecycleSnapshot(phase: ${phase.name}, '
      'wakePending: $wakePending, closeRequested: $closeRequested, '
      'cancelRequested: $cancelRequested, permitHeld: $permitHeld)';
}

final class SyncLifecycleTransitionError implements Exception {
  const SyncLifecycleTransitionError(this.before, this.event);

  final SyncLifecycleSnapshot before;
  final SyncLifecycleEvent event;

  @override
  String toString() =>
      'undefined opto-sync lifecycle transition: '
      '${before.phase.name} + ${event.name} ($before)';
}

/// Small deterministic machine used directly by production runners.
///
/// Undefined events never guess a recovery path: [apply] throws and leaves the
/// state unchanged. Public idempotence (for example, calling `close` twice) is
/// handled at the API boundary before entering this machine.
final class SyncLifecycleMachine {
  SyncLifecycleSnapshot _state = const SyncLifecycleSnapshot.initial();

  SyncLifecycleSnapshot get state => _state;

  bool allows(SyncLifecycleEvent event) => _next(_state, event) != null;

  SyncLifecycleSnapshot apply(SyncLifecycleEvent event) {
    final next = _next(_state, event);
    if (next == null) throw SyncLifecycleTransitionError(_state, event);
    if (!next.isValid) {
      throw StateError('lifecycle transition produced an invalid state: $next');
    }
    _state = next;
    return next;
  }

  static SyncLifecycleSnapshot? transition(
    SyncLifecycleSnapshot state,
    SyncLifecycleEvent event,
  ) {
    if (!state.isValid) return null;
    return _next(state, event);
  }

  static SyncLifecycleSnapshot? _next(
    SyncLifecycleSnapshot state,
    SyncLifecycleEvent event,
  ) {
    switch (event) {
      case SyncLifecycleEvent.wake:
        if (state.phase == SyncLifecyclePhase.closed || state.closeRequested) {
          return null;
        }
        return state.copyWith(wakePending: true);
      case SyncLifecycleEvent.join:
        if (state.phase != SyncLifecyclePhase.acquiring &&
            state.phase != SyncLifecyclePhase.running &&
            state.phase != SyncLifecyclePhase.releasing) {
          return null;
        }
        return state;
      case SyncLifecycleEvent.beginAcquire:
        if (state.phase != SyncLifecyclePhase.idle ||
            !state.wakePending ||
            state.closeRequested) {
          return null;
        }
        return state.copyWith(
          phase: SyncLifecyclePhase.acquiring,
          wakePending: false,
        );
      case SyncLifecycleEvent.acquireGranted:
        if (state.phase != SyncLifecyclePhase.acquiring) return null;
        return state.copyWith(
          phase: state.closeRequested
              ? SyncLifecyclePhase.releasing
              : SyncLifecyclePhase.running,
          permitHeld: true,
          cancelRequested: false,
        );
      case SyncLifecycleEvent.acquireDeferred:
        if (state.phase != SyncLifecyclePhase.acquiring) return null;
        return state.copyWith(
          phase: state.closeRequested
              ? SyncLifecyclePhase.closed
              : SyncLifecyclePhase.idle,
          wakePending: state.closeRequested ? false : state.wakePending,
          cancelRequested: false,
          permitHeld: false,
        );
      case SyncLifecycleEvent.cancel:
        if (state.phase != SyncLifecyclePhase.running) return null;
        return state.copyWith(cancelRequested: true);
      case SyncLifecycleEvent.cycleSettled:
        if (state.phase != SyncLifecyclePhase.running || !state.permitHeld) {
          return null;
        }
        return state.copyWith(
          phase: SyncLifecyclePhase.releasing,
          cancelRequested: false,
        );
      case SyncLifecycleEvent.releaseSettled:
        if (state.phase != SyncLifecyclePhase.releasing || !state.permitHeld) {
          return null;
        }
        return state.copyWith(
          phase: state.closeRequested
              ? SyncLifecyclePhase.closed
              : SyncLifecyclePhase.idle,
          wakePending: state.closeRequested ? false : state.wakePending,
          cancelRequested: false,
          permitHeld: false,
        );
      case SyncLifecycleEvent.close:
        if (state.phase == SyncLifecyclePhase.closed) return null;
        if (state.phase == SyncLifecyclePhase.idle) {
          return state.copyWith(
            phase: SyncLifecyclePhase.closed,
            wakePending: false,
            closeRequested: true,
            cancelRequested: false,
          );
        }
        return state.copyWith(
          wakePending: false,
          closeRequested: true,
          cancelRequested: state.phase == SyncLifecyclePhase.running,
        );
      case SyncLifecycleEvent.processAbort:
        if (state.phase != SyncLifecyclePhase.acquiring &&
            state.phase != SyncLifecyclePhase.running &&
            state.phase != SyncLifecyclePhase.releasing) {
          return null;
        }
        return state.copyWith(
          phase: state.closeRequested
              ? SyncLifecyclePhase.closed
              : SyncLifecyclePhase.idle,
          wakePending: false,
          cancelRequested: false,
          permitHeld: false,
        );
    }
  }
}
